### ISSUE TITLE
How to Ignore 3rd party exe while doing code sign

### ISSUE THREAD

**Issue by ramkirsr** [electron-userland/electron-builder#6304]
* **Electron-Builder Version**: 22.11.7

* **Node Version**: v14.18.0

* **Electron Version**:v11.5.0
* **Electron Type (current, beta, nightly)**:current

* **Target**: windows

We are creating a electron app using the electron zoom sdk, when we bundle the app using electron builder its sign all the dependency exe files. Because of which zoom sdk is not working properly.

I have tried manual sign -> electron builder not allowing to do auto update, throwing the signature miss match error

Basically,
1)  I want to know, is there a way to skip folder sign?

2) What is the format which electron builder will use for code sign?, so that i can do manual sign and release the build.

**Comment by ramkirsr** [comment]
Please update us, its quite critical.

**Comment by mmaietta** [comment]
Looks like you should be able to set `signIgnore: [array of regexes]` to 
```
mac: {
   signIgnore: ["\.exe"]
}
```
https://github.com/electron-userland/electron-builder/blob/8ebfc96276bffe0bc1ad394c5ae6843976e01709/packages/app-builder-lib/src/macPackager.ts#L230-L239

Haven't tried this myself though

**Comment by ramkirsr** [comment]
> Looks like you should be able to set `signIgnore: [array of regexes]` to
> 
> ```
> mac: {
>    signIgnore: ["\.exe"]
> }
> ```
> 
> https://github.com/electron-userland/electron-builder/blob/8ebfc96276bffe0bc1ad394c5ae6843976e01709/packages/app-builder-lib/src/macPackager.ts#L230-L239
> 
> Haven't tried this myself though

--> Thanks for your kind support , In my use case am generating the build for Windows.

**Comment by ramkirsr** [comment]
@mmaietta  Thanks for your kind ### support , In my use case i am generating the build for Windows.

**Comment by mmaietta** [comment]
Ah, my apologies, not sure how I misread that.

Well, looking at `signtool.exe`, it doesn't have any built-in manner with which to exclude specific files.
https://docs.microsoft.com/en-us/dotnet/framework/tools/signtool-exe

You'll probably need to override windows signing and use your own custom signing mechanism/process via:
```
win: {
   sign: async (config: WindowsSignTaskConfiguration) => {

   }
}
```

**Comment by ramkirsr** [comment]
@mmaietta ; Thank a lot for your kind response.
I have tried using signtool to sign the exe, but when i have tried to push the upgrade, signature miss match error happened and upgrade failing. Do we have anything like, what is the sign params is used by electron builder?   and how the they are verifying the signature at the time of upgrade?

**Comment by mmaietta** [comment]
This is the signing arg compilation
https://github.com/electron-userland/electron-builder/blob/7f933d0004a0a5f808a2a1c71dca7362cab2728e/packages/app-builder-lib/src/codeSign/windowsCodeSign.ts#L188-L208

Signature verification cmd:
https://github.com/electron-userland/electron-builder/blob/7f933d0004a0a5f808a2a1c71dca7362cab2728e/packages/electron-updater/src/windowsExecutableCodeSignatureVerifier.ts#L42

**Comment by ramkirsr** [comment]
@mmaietta ; I have written the custom script to sign only my exe; verify the code is it proper. 

const signtool = "\"C:/Program Files (x86)/Windows Kits/10/bin/10.0.19041.0/x64/signtool.exe\""

    if(Path.basename(configuration.path).toLowerCase().startsWith("tribyte") && configuration.hash == "sha1") {

        require("child_process").execSync(
            `${signtool} sign /${configuration.hash} ${configuration.sha1_thumbprint} /v /f "${configuration.cscInfo.file}" /p ${configuration.cscInfo.password} /fd ${configuration.hash} /t http://timestamp.digicert.com /td ${configuration.hash} "${configuration.path}"`,
            {
                stdio: "inherit"
            }
        );
    }
    else if(Path.basename(configuration.path).toLowerCase().startsWith("tribyte") && configuration.hash == "sha256") {
      
        require("child_process").execSync(
            `${signtool} sign /sha1 ${configuration.sha1_thumbprint} /as /v /f "${configuration.cscInfo.file}" /p ${configuration.cscInfo.password} /fd ${configuration.hash} /tr http://timestamp.comodoca.com/rfc3161 /td ${configuration.hash} "${configuration.path}"`,
            {
                stdio: "inherit"
            }
        );
    }

**Comment by minht11** [comment]
For anyone still looking how to exclude some files from Windows signing while not using custom sign tool here is a custom script. `WIN_CSC_LINK` and `WIN_CSC_KEY_PASSWORD` need to be set as env variables.
```js
const path = require('path')
const { doSign } = require('app-builder-lib/out/codeSign/windowsCodeSign')

/**
 * @type {import("electron-builder").CustomWindowsSign} sign
 */
module.exports = async function sign(config, packager) {
  // Do not sign if no certificate is provided.
  if (!config.cscInfo) {
    return
  }

  const targetPath = config.path

  const ext = path.extname(targetPath)
  const baseName = path.basename(targetPath, ext)

  // Do not sign elavate file, because that prompts virus warning.
  if (targetPath.endsWith('elevate.exe')) {
    return
  }

  // Only sign our own DLL files.
  if (ext === '.dll' && !baseName.toLowerCase().startsWith('mylib')) {
    return
  }

  await doSign(config, packager)
}
```

**Comment by serg06** [comment]
@minht11 Can you please explain how to connect that script to the build process?

**Comment by mmaietta** [comment]
@serg06 in your electron-builder config
```
{
  win: {
    sign: "path to sign.js"
  }
}
```
Ref: `sign` https://www.electron.build/configuration/win
